tumcms/Open-Infra-Platform
# Introducing inverse

First, #195 should be merged. This Pull Request builds on top - all changes with paths starting with `UnitTests` can be safely ignored.

Fixes #35 .

**tl;dr:**

INVERSE parameters are now parsed and filled with data when reading.

**What has been done:**
- `Entity` has a new function `setInverseCounterpart` that links this->attribute with the other->attribute
- `EntityAttribute` now knows about its inverse status, as well as about the inverse's counterpart's name (only the true inverse has its `inverse` flag set to true)
- `Schema::linkInverses()` links all inverse attributes among eachother
- parser now understands the inverse statements
- generator was adapted accordingly:
  - each entity now has a function `void linkInverse()` that links itself to the inverse attribute of all attributes with inverse counterparts (`this->attr->inverseAttribute->add(this)`
  - `main` links the inverses just before generating the files

**How to test?**

Using the UT_schema_v3:
- both `IfcAlignment2dHorizontal` and `-Vertical` have an inverse attribute `ToAlignmentCurve FOR IfcAlignmentCurve` 
- when generated, both receive an inverse attribute `SET<IfcAlignmentCurve> ToAlignmentCurve`
- `IfcAlignmentCurve::linkInverse` links itself to the corresponding inverse attribute of `-Horizontal` and `-Vertical`

Build sequence:
- build `OpenInfraPlatform.ExpressBindingGenerator`
- build `Generate.schema_UT_v3`
- see the generated files (`UnitTests\Generator\schema_UT_v3\Generated`)
